File: c/cz/mg/vulkan/PFN_vkQueueSubmit.c
#include <vulkan/vulkan.h>
#include <jni.h>
#include <stdlib.h>
#include <string.h>

jlong jniPointerToLong(void* p);
void* jniLongToPointer(jlong l);
jlong jniFunctionPointerToLong(PFN_vkVoidFunction p);
PFN_vkVoidFunction jniLongToFunctionPointer(jlong l);
void jniThrowException(JNIEnv* env, const char* message);

void Java_cz_mg_vulkan_vk_PFNvkQueueSubmit_call(JNIEnv* env, jclass jc, jlong address, jlong queue, jlong submitCount, jlong pSubmits, jlong fence, jlong rval){
    (void)env;
    (void)jc;
    PFN_vkQueueSubmit f = (PFN_vkQueueSubmit)jniLongToFunctionPointer(address);
    VkResult* rvalAddress = jniLongToPointer(rval);
    *rvalAddress = f(
        *((VkQueue*)jniLongToPointer(queue)),
        *((uint32_t*)jniLongToPointer(submitCount)),
        ((VkSubmitInfo*)jniLongToPointer(pSubmits)),
        *((VkFence*)jniLongToPointer(fence))
    );
}

File: cz/mg/vulkan/PFNvkQueueSubmit.java
package cz.mg.vulkan;

public class PFNvkQueueSubmit extends VkFunctionPointer {
    public PFNvkQueueSubmit() {
    }

    public PFNvkQueueSubmit(VkMemory vkmemory) {
        super(vkmemory);
    }

    public PFNvkQueueSubmit(VkMemory vkmemory, long vkaddress) {
        super(vkmemory, vkaddress);
    }



    public PFNvkQueueSubmit(long value) {
        setValue(value);
    }

    public PFNvkQueueSubmit(VkInstance instance) {
        super(instance, new VkString("vkQueueSubmit"));
    }

    public void call(VkQueue queue, VkUInt32 submitCount, VkSubmitInfo pSubmits, VkFence fence, VkResult rval){
        call(getValue(), queue != null ? queue.getVkAddress() : VkPointer.getSinkAddress(), submitCount != null ? submitCount.getVkAddress() : VkPointer.getSinkAddress(), pSubmits != null ? pSubmits.getVkAddress() : VkPointer.NULL, fence != null ? fence.getVkAddress() : VkPointer.getSinkAddress(), rval != null ? rval.getVkAddress() : VkPointer.getSinkAddress());
    }

    private static native void call(long vkaddress, long queue, long submitCount, long pSubmits, long fence, long rval);
}
